feat(forms): add `moveControl(from, to)` to `FormArray`

ngfelixl opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)

Not sure about the next one, since this is my first PR in this repo. Is it meant to update the public_api_guard? In the *packages/forms/PACKAGE.md* there is no section regarding FormArrays.

- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Particularly for using cdk drag-drop, there should be a function that savely moves an item in a FormArray, but there is none and one has to reorder the items manually. It should be similar to the CDK `moveItemInArray()` function provided for Angular 7 cdk drag and drop package.

Issue Number: N/A


## What is the new behavior?

With this patch one can create a formarray, say `a` with 4 elements having formcontrols with values [0, 1, 2, 3] and use `a.moveItem(0, 2)` which moves the item at position 0 to position 2 in the formarray results in [1, 2, 0, 3]. It throws an error using the private `_throwIfControlMissing`, if one of the items in between does not exist. Also works with nested forms.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information

Have opened a feature request before in this post https://github.com/angular/angular/issues/27171.

Airblader commented:
I think there should be more edge case tests such as moving an item out of bounds or trying to move from an index that is out of bounds (I know you implemented it, but there aren't tests for them).

kara commented on packages/forms/src/model.ts:
```diff
@@ -1715,6 +1715,27 @@ export class FormArray extends AbstractControl {
+    this._throwIfControlMissing(to);
+
+    const delta = to > from ? 1 : -1;
+    const temp = this.at(from);
```
Nit: Naming could be more specific

```suggestion
    const controlToMove = this.at(from);
```

kara commented on packages/forms/src/model.ts:
```diff
@@ -1715,6 +1715,27 @@ export class FormArray extends AbstractControl {
+      this.setControl(i, this.at(i + delta));
+    }
+
+    this.setControl(to, temp);
```
We should have a sanity check that `from` is not equal to `to`. If it is, we should do nothing. Can you also add a test for that?

ngfelixl replied:
Sure. Implemented the check and appropriate tests. Included your other suggestions as well. Currently running the test suite and will submit the changes when the test pass.

kara commented on packages/forms/src/model.ts:
```diff
@@ -1715,6 +1715,27 @@ export class FormArray extends AbstractControl {
+   * @param from Index of the array item to move
+   * @param to Index of the target item position
+   */
+  moveItem(from: number, to: number): void {
```
How about this?

```suggestion
  moveControl(from: number, to: number): void {
```

ngfelixl commented:
I guess something went wrong while rebasing. Followed along the [developers guide](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md). I am sorry for the mess, any suggestions how to fix it?

kara commented on packages/forms/src/model.ts:
```diff
@@ -1736,6 +1736,31 @@ export class FormArray extends AbstractControl {
   }
 
+  /**
+   * Move a control from one position to another
```
Since this is a new public API, can we add some more detailed docs?  I think an example here of how to use this method would be helpful, and we should probably add some more details to the commit message as well.

ngfelixl replied:
Thanks for your review and advices. Added more documentation including examples, squashed the commits and improved the commit message.

I don't know why the **ci/circleci: legacy-unit-tests-saucelabs** test is not passing. It just says "couldn't connect to https://saucelabs.com/rest/v1/angular-ci/tunnels". Any suggestions?

kara reviewed (changes requested):
Sorry to hear about the rebase issues !  Would you mind squashing your commits into one?

```bash
git reset HEAD~4
git add --all
git commit
git push origin formarray-moveitem -f
```

ngfelixl committed:
feat(forms): implement moveControl(from, to) formarray memberfunction

Implement moveControl(from: number, to: number) FormArray
memberfunction. It moves a control to another position in the
FormArray. This can be particularly useful in combination with
Material CDK drag and drop.

kara commented on packages/forms/src/model.ts:
````diff
@@ -1736,6 +1736,78 @@ export class FormArray extends AbstractControl {
+   * console.log(arr.value);   // [1, 2, 0]
+   * ```
+   *
+   * ### Use with Material CDK drag and drop
````
Do you mind removing this section? We probably shouldn't have docs for Material components/utilities in core (if API changes, these will easily get stale). Maybe link to material docs instead?

> This can be used with Material CDK drag and drop. See its docs here:  <link>

ngfelixl replied:
That makes sense :+1: Revised it

kara reviewed (commented):
Thanks for adding docs! They are great; just one comment and we should be good to go.

kara reviewed (approved):
LGTM

[presubmit](http://test/OCL:236209238:BASE:236209271:1551396608704:8161f26d)

kara commented:
@IgorMinar Can you take a look for public-api approval?

skrtheboss commented:
Hi, i wanted to ask if there is a reason why you call `setControl` instead of just moving the items on the `controls` array. By calling `setControl` `valueChanges` also emits for each iteration, so it results in many value changes.

I would suggest to use one of these implementations:

[@angular/cdk `moveItemInArray` like implementation](https://github.com/angular/material2/blob/master/src/cdk/drag-drop/drag-utils.ts#L15)
```
public moveControl(from: number, to: number): void {
    this._throwIfControlMissing(from);
    this._throwIfControlMissing(to);

    if(from === to) {
        return;
    }

    const delta = to > from ? 1 : -1;
    const controlToMove = this.controls[from];

    for(let i = from; i !== to; i += delta) {
        this.controls[i] = this.controls[i + delta];
    }

    this.controls[to] = controlToMove;

    this.updateValueAndValidity();
    this._onCollectionChange();
}

``` 
[As suggested on stack-overflow](https://stackoverflow.com/questions/5306680/move-an-array-element-from-one-array-position-to-another/5306832#5306832)
```
public moveControl(from: number, to: number): void {
    this._throwIfControlMissing(from);
    this._throwIfControlMissing(to);

    if(from === to) {
        return;
    }

    this.controls.splice(to, 0, this.controls.splice(from, 1)[0]);

    this.updateValueAndValidity();
    this._onCollectionChange();
}
```

ngfelixl commented:
@kara @petebacondarwin Are there any suggestions related to previous comment? Do you still would like to have this feature and should I update it to @skrtheboss suggestions?

skrtheboss commented:
@ngfelixl at our company we are currently using a extended version of the angular `FormArray`, and it is working flawlessly, can you have a look?
```
export class EnhancedFormArray extends FormArray {

    /**
     * Move a control from one position to another. The controls
     * in between move one position towards the moved controls
     * origin.
     *
     * @usageNotes
     * ### Move a control inside of a FormArray
     *
     * ```typescript
     * const arr = new FormArray([
     *   new FormControl(0),
     *   new FormControl(1),
     *   new FormControl(2)
     * ]);
     *
     * arr.moveControl(0, 2);
     * console.log(arr.value);   // [1, 2, 0]
     * ```
     *
     * This can be used with Material CDK drag and drop. See its
     * docs here: https://material.angular.io/cdk/drag-drop/overview
     *
     * @param from Index of the array item to move
     * @param to Index of the target item position
     */
    public moveControl(from: number, to: number): void {
        (this as any)._throwIfControlMissing(from);
        (this as any)._throwIfControlMissing(to);

        if(from === to) {
            return;
        }

        this.controls.splice(to, 0, this.controls.splice(from, 1)[0]);

        this.updateValueAndValidity();
        (this as any)._onCollectionChange();
    }

    /**
     * Replace the current controls with new ones.
     *
     * @usageNotes
     * ### Replace controls of a FormArray
     *
     * ```typescript
     * const arr = new FormArray([
     *   new FormControl(0),
     *   new FormControl(1),
     *   new FormControl(2)
     * ]);
     *
     * console.log(arr.value);   // [1, 2, 0]
     *
     * arr.replaceControls([new FormControl(2)]);
     *
     * console.log(arr.value);   // [2]
     * ```
     *
     * @param controls The new controls which will replace the old ones
     */
    public replaceControls(controls: AbstractControl[]): void {
        if(this.controls.length) {
            this.controls.forEach(control => (control as any)._registerOnCollectionChange(() => undefined));
            this.controls.splice(0);
        }

        controls.forEach(control => {
            this.controls.push(control);
            (this as any)._registerControl(control);
        });

        this.updateValueAndValidity();
        (this as any)._onCollectionChange();
    }
}

```

petebacondarwin commented:
I think I agree that @skrtheboss's approach is more performant.
Could you try this but add in some tests to ensure that the correct outputs are emitted when the move occurs?

ngfelixl committed:
refactor(forms): update form state only after move

Remove and add controls via splice function
Update the forms value and validity after move
Trigger _onCollectionChange after move

ngfelixl commented:
@petebacondarwin I changed the code to agreed suggestion and it seems to work accordingly and the existing tests pass. Do you think I have to add even more tests? Currently I don't see any uncovered edge-case related to the second implementations.

petebacondarwin commented:
@ngfelixl - thanks for making these changes. Regarding tests, I would like to see some that check what is happening with the change events. For example there is this test for `setControl()` : https://github.com/angular/angular/blob/fb7ec087f313d73d624029d44f1f16e610597c3d/packages/forms/test/form_array_spec.ts#L1297-L1303

It would be good to have some similar tests for `moveControl()`. Perhaps adding them to this describe block: https://github.com/angular/angular/blob/fb7ec087f313d73d624029d44f1f16e610597c3d/packages/forms/test/form_array_spec.ts#L811.

Also it would be good to have tests that make it clear what happens to `isDirty/isPristine` when a control is moved; and also check that things like validation work as expected.

ngfelixl commented:
Thanks for clarification. I will implement additional tests!

ngfelixl commented:
@petebacondarwin one question regarding the states. From within the doc comments I would think of the following

- **dirty**: it should not be affected because it says *"Programmatic changes to a control's value do not mark it dirty."*. Applying the moveControl function is a programmatic change. Same goes for pristine.
- **touched**: it should not be affected because it says *"A control is marked touched once the user has triggered a blur event on it."*.

So I will implement tests that will check that the prior state (before movement) does not change with a moveControl execution. Do you agree?

petebacondarwin commented:
That sounds very reasonable. Thanks!

ngfelixl committed:
test(forms): add validation, valuechanges, dirty/touch tests for move

Implement test checking form validation after movement
Implement test checking for independent state behavior
(dirty/touched)
Implement tests checking correct valueChanges behavior
for correct value and 1 emit/per move

ngfelixl commented:
@petebacondarwin I've added more tests. I don't know why the integration test fails. It passed before the latest commit. The error message seems to be unrelated to the formarray. Maybe try a rerun?

petebacondarwin commented:
Integration test failure is unrelated - see https://github.com/angular/angular/pull/34179

petebacondarwin commented on packages/forms/test/form_array_spec.ts:
```diff
@@ -638,6 +709,23 @@ import {of } from 'rxjs';
+        g.moveControl(0, 1);
+        expect(g.valid).toEqual(false);
+        expect(g.errors).toEqual({'broken': true});
+      });
```
Love this test!

ngfelixl replied:
Awesome, thank you for reviewing! Yeah, I agree. This test was a lot of fun to implement.

What are the next steps now, can I do anything to support merge process? Do you have any other tasks/open issues where I can support?

petebacondarwin replied:
The next step are to get it approved for public API change, and to get it through Google3 internal presubmit tests. Only a Googler can do these. I have pinged them to help.

skrtheboss commented:
@ngfelixl if you have some time left, is there any chance you could also implement a `replaceControls` method? Should i open a new issue for that?

Here is my implementation: 
```
    /**
     * Replace the current controls with new ones.
     *
     * @usageNotes
     * ### Replace controls of a FormArray
     *
     * ```typescript
     * const arr = new FormArray([
     *   new FormControl(0),
     *   new FormControl(1),
     *   new FormControl(2)
     * ]);
     *
     * console.log(arr.value);   // [1, 2, 0]
     *
     * arr.replaceControls([new FormControl(2)]);
     *
     * console.log(arr.value);   // [2]
     * ```
     *
     * @param controls The new controls which will replace the old ones
     */
    public replaceControls(controls: AbstractControl[]): void {
        if(this.controls.length) {
            this.controls.forEach(control => (control as any)._registerOnCollectionChange(() => undefined));
            this.controls.splice(0);
        }

        controls.forEach(control => {
            this.controls.push(control);
            (this as any)._registerControl(control);
        });

        this.updateValueAndValidity();
        (this as any)._onCollectionChange();
    }
```

I need it if i make custom `ControlValueAccessor`s with an array as input in the `wtiteValue` function, to replace controls with the new array.

petebacondarwin commented:
That should be a new PR if you implement it.

ngfelixl commented:
@skrtheboss sure I would like to implement that one. Can you please open an issue with the reason of the implementation and ~ how to use it. And mark me as "willing to implement" or developer? Then I open another PR and link that to your issue. @petebacondarwin should I await an "ok" from a member before I start creating a PR?

petebacondarwin commented:
I cannot guarantee that it will be accepted but it looks like a reasonable feature to me.

skrtheboss commented:
@ngfelixl is opened a new [issue](https://github.com/angular/angular/issues/34186). Thank you.

petebacondarwin commented:
@ngfelixl - sorry this has been sitting in the backlog for so long. If you are still interested in getting it merged, could you rebase and fix the merge conflicts? Then hopefully we can get it presubmitted and see where we can land it.

ngfelixl commented:
@petebacondarwin thank you :) yes, of course, I would be very happy to complete this! I will work on it tomorrow.

petebacondarwin commented:
@ngfelixl - before you start, @AndrewKushnir had a good idea about making this feature tree shakable. His idea is to make this a free-standing helper function (e.g. `moveControl(arr, from, to)`) rather than a method.

ngfelixl commented:
@petebacondarwin @AndrewKushnir sounds good, I've got a few questions then. Where should it be located best and how should it be exported? Would it be available in '@angular/forms'?

petebacondarwin commented:
It would be exported from `@angular/forms`, located alongside alongside the `FormArray`, etc, classes.

petebacondarwin commented:
Although forms doesn't yet have any helper functions, there are plenty of examples in other packages: https://angular.io/api?type=function

AndrewKushnir commented:
Hi @ngfelixl, thanks for creating this PR!

As Pete mentioned, I've created a ticket that describes the problem (extra non-tree-shakable code) and a possible solution, see #39640. I believe we should do more research on our end before we can proceed with adding separate util methods as this approach deviates from the current path (where the logic is added to the AbstractControl-based classes) and it would potentially make Forms API inconsistent.

Another tricky aspect is that calling current API methods (to remove a control from one location and append it in a new location) would trigger 2 `updateValueAndValidity` calls, thus causing validators to run twice and potentially cause more unnecessary actions that would be invalidated by the subsequent `updateValueAndValidity` call.

I'd propose to put the work on the `moveControl` helper on hold until there is a good plan in place.

Thank you.

ngfelixl commented:
Hi @AndrewKushnir, thank you for your insights. It makes all sense to me. I won't start implementing any changes until you give a "good-to-go". If there is anything I can do in order to help or/and support you in the meanwhile, please let me know.

AndrewKushnir commented:
@ngfelixl thanks for the reply! We'll let you know once we get back to the tree-shaking question for Forms package (which should unblock this and similar PRs).

AndrewKushnir commented:
FYI, I'm going to close this PR for now as it's not actionable (we can not merge it) without better tree-shaking support in Forms package, see #39640. This PR is linked in #39640, so we'll re-consider once there is a way to add helper methods in a tree-shakable way.
